### DISCUSSION TITLE
Engine lags when you add more than one menu

### DISCUSSION THREAD

**Discussion by Raptorq1** [Techokami/SonicWorldsNext#149] [Q&A]
Engine lags when you add more than one menu

**Comment by r6915ee** [comment]
As in, whenever you add a menu on top of another, the frame rate drops?

**Reply by Raptorq1** [reply]
Yes but drops at random times

**Reply by r6915ee** [reply]
What are some of the most common times it does drop?

**Reply by Raptorq1** [reply]
I don't know

**Reply by r6915ee** [reply]
What about selecting a specific option then?

**Reply by Raptorq1** [reply]
Yes

**Reply by r6915ee** [reply]
Well then, what are some common types of selecting an option causing framerate? Like, for example, if it was some sort of Zone Select menu, then the option to go to one of the Zones?

**Reply by Raptorq1** [reply]
My game has a auto save waning when I select it and then select a zone it will lag

**Reply by r6915ee** [reply]
Maybe it could be something in the scripts used for the auto save warning or selecting a zone? You might be probably not removing the warning in your script after it's gone.

**Reply by Raptorq1** [reply]
It's still happening

**Reply by r6915ee** [reply]
Does it either jitter or stutter, [according to Godot's definition?](https://docs.godotengine.org/en/stable/tutorials/rendering/jitter_stutter.html)

**Reply by Raptorq1** [reply]
I don't know oh and when it happens I hear 4 jumping sfx all at once and entering a special ring seems to fix it but it makes you lose a life 
![Screenshot_20240613-235354](https://github.com/Techokami/SonicWorldsNext/assets/106418027/5a42ed4f-7710-410b-af3d-68b483767033)

**Reply by r6915ee** [reply]
The 4 jumping SFX likely accumulates to your discussion about splitscreen. Also, what do you exactly mean by the latter half in specific?

**Reply by Raptorq1** [reply]
No the split screen thing is separate Here's a screenshot of the original chat 
![Screenshot_20240614-003454](https://github.com/Techokami/SonicWorldsNext/assets/106418027/ea736849-3581-4c1e-bf48-02ce66f3182c)
Oh and when you try entering a special ring after its fix it doesn't give you a chaos Emerald

**Reply by r6915ee** [reply]
Four players jumping at once roughly accumulates to your splitscreen, so in some way it might also be the same. I'd suggest following DimensionWarped's advice for both problems as well. The Godot output console should also tell you if there are any problems at times.

**Reply by Raptorq1** [reply]
Iam on mobile I can't do that

**Reply by r6915ee** [reply]
Then can't you just put the project onto any sort of computer, and run it on the version of Godot on there? I don't know why you would make it on mobile in the first place.

**Reply by Techokami** [reply]
Because he doesn't HAVE a computer.

**Reply by r6915ee** [reply]
Well then, I guess maybe see what code runs when you enter a zone? If it's SPECIFICALLY whenever there is certain interaction with the auto-save warning, then it might be best to check that as well.

**Comment by DimensionWarped** [comment]
This is strictly a you problem my dude and there is no way anyone could ever hope to solve it with that description.

**Reply by Raptorq1** [reply]
Hey you know that I had this problem before when I first started making my fangame

**Reply by r6915ee** [reply]
> Hey you know that I had this problem before when I first started making my fangame

Then how come you didn't ask about it in the first place, though?

**Reply by r6915ee** [reply]
And if it's right when you began using the engine, but didn't develop anything, that could either be your device or just the code being unoptimized for your device.

**Reply by Raptorq1** [reply]
You could be right My device is a Amazon fire 10 HD that could be it I am saying this because of this Generally, Android is stutter and jitter-free because the running activity gets all the priority. That said, there may be problematic devices (older Kindle Fire is known to be one). If you see this problem on Android, please let us know.

**Reply by r6915ee** [reply]
I mean, Amazon Fire 10 HD did come out about a decade ago (September 2015). Since a Fire 10 HD should be part of the Kindle Fire lineup, it likely falls into that category as well.

**Reply by Raptorq1** [reply]
Do you have any ideas on how to fix it or should I put my fan game on hold until I get a laptop

**Reply by r6915ee** [reply]
I guess for JUST the lag alone it would be best to leave that issue for now, and perhaps have any playtesters on the project see if there are any issues on their end. Getting a laptop I'd say would also increase your scope of figuring out these issues, testing out if this issue still persists, and is generally more flexible, though with the rate of progression I can accumulate based on some of the things you shared I wouldn't say you should put it on hold.

**Reply by Raptorq1** [reply]
Okay but how would I do the play test thing if I can't export the project to pc or Android

**Reply by r6915ee** [reply]
Then a possible solution would just be to possibly hand the project files to the playtesters, since it should be available somewhere in the storage. Of course, this comes with the risk of possible leaks of the fangame, but it's kind of the last resort in a way for this situation.

**Reply by Raptorq1** [reply]
Okay but the other thing the play testers would have to get godot 4

**Reply by r6915ee** [reply]
Generally it shouldn't be much of an issue. Either one coder on the fangame who has access to exporting could export builds of the game for the playtesters, the playtesters can use their version of Godot to test the project, or the Godot web editor (possibly, I don't use it).

**Reply by DimensionWarped** [reply]
> Do you have any ideas on how to fix it or should I put my fan game on hold until I get a laptop

You could always just work on other things and ignore the little bit of menu lag... that said, I still can't imagine how anyone could efficiently work on a fangame without a proper computer.

In any event, best way to get help on this would be to post a video. A menu is a weird place for stuff to be lagging if it isn't lagging without menus.

**Reply by Raptorq1** [reply]
Okay here's video proof 

https://github.com/Techokami/SonicWorldsNext/assets/106418027/45473aa9-a43d-409c-83ab-0cd4aa4cee1a


https://github.com/Techokami/SonicWorldsNext/assets/106418027/5a17a335-d3a1-4457-a551-4c438ad92563


And when in a special stage while the lag is active and you touch a checkpoint the game crashes

**Reply by DimensionWarped** [reply]
Uh... first, that's not the standard menu that comes with the game right now, so you have gone into the territory of asking us to psychically know what you've done to the engine...

second, the lag appears to start *before* you open the menu in this video and there are multiple Sonic sprites on the screen so I'm far more inclined to think that something you did with Super Sonic is responsible for this lag than the menus.

**Reply by Raptorq1** [reply]
I never messed with the super sonic scripts well just to change the music the lag was happening because of the menus When I went back to the title screen it would fade twice

**Reply by DimensionWarped** [reply]
Funny, I've never seen Sonic clone himself when turning Super Sonic in the stock engine.

**Reply by Raptorq1** [reply]
It doesn't only happen when turning super but when he's normal sonic as well

**Reply by Raptorq1** [reply]
Oh if I add more menus to the  unmodified engine it will start to lag

**Reply by DimensionWarped** [reply]
How about you demonstrate that by providing a minimal patch to reproduce the problem so that we can actually determine what your problem is.

**Reply by Raptorq1** [reply]
> How about you demonstrate that by providing a minimal patch to reproduce the problem so that we can actually determine what your problem is.

I will give you the fork via discord

**Reply by DimensionWarped** [reply]
I don't need you to be playing secret squirrel and sending this stuff just to me.

https://github.com/Raptorq1/SonicWorldsNexttest

So I looked this over and it doesn't reproduce the issue you are trying to reproduce. Instead it just adds a bunch of copies of the character/level select screen scene and scripts, accidentally deleting the only CharacterSelect.gd that actually would have been read resulting in an unusable character select screen.

Renaming any one of those character select scripts to CharacterSelect.gd makes the character select screen work again and it doesn't appear any of the other copies get touched by the engine when I do, so I'm really not sure why you thought this would change anything.

**Reply by Raptorq1** [reply]
Okay then its an issue with my device

**Reply by DimensionWarped** [reply]
If you are truly working off the build you linked, then it's a matter of how you changed it, not of your device.

**Reply by Raptorq1** [reply]
I might ask godot forums for help to see what they say

**Comment by Renhoex** [comment]
I'm going to be blunt, we need more information about your issue or we can't help you because we have no idea if this is a serious problem with our framework or if what you're specifically doing is causing issues with out framework.

I'll just outline some questions if you could answer them clearly we *might* be able to at least guide you.

1. did you alter the original pause menu script or change any of the nodes? if you did what did you change?
2. What activated your menu? is it pressing the pause button, did you disable the regular pausing routine so your menu can work?
3. are you sure the menu is disabled when you pause
4. are you creating your new menu through instantiate? If so, did you set it to queue free when it's closed? If not your game could be lagging because you're creating too many menu instances or your menu is still saving causing your game to lag.

**Reply by Raptorq1** [reply]
No I didn't Chang anything from the pause menu
It is a button pressed but the menu item is only accessed before you go to the character select menu
yes
No

**Reply by NeonSRB2** [reply]
Is it a seperate scene from the Character Select, or part of the title screen?
Also does it lag during gameplay?

**Reply by Raptorq1** [reply]
The menus are all separate oh and the character select isnt being used for my game and it happens at random times and when the lag does happen it happens in game as well
